torchtune/utils/_distributed.py

```diff
@@ -242,6 +250,83 @@ def lora_wrap_fsdp(module: nn.Module, recurse: bool, **kwargs):
     return lora_wrap_fsdp
 
 
+def load_from_full_model_state_dict(
+    model: FSDPModule,
+    full_sd: Dict[str, Any],
+    device: torch.device,
+    is_rank_zero: bool,
+):
+    meta_sharded_sd = model.state_dict()
+    sharded_sd = {}
+    for param_name, full_tensor in full_sd.items():
+        sharded_meta_param = meta_sharded_sd.get(param_name)
+        # `.to(dtype)` ensures same dtype when `assign=True`
+        sharded_tensor = distribute_tensor(
+            full_tensor.to(sharded_meta_param.dtype),
+            sharded_meta_param.device_mesh,
+            sharded_meta_param.placements,
+        )
+        sharded_sd[param_name] = nn.Parameter(sharded_tensor)
+    # choose `assign=True` since we cannot call `copy_` on meta tensor
+    model.load_state_dict(sharded_sd, strict=False, assign=True)
+
+
+def get_full_model_state_dict(
+    model: FSDPModule,
+    is_rank_zero: bool,
+) -> Dict[str, Any]:
+    sharded_sd = model.state_dict()
+    cpu_state_dict = {}
+    for param_name, sharded_param in sharded_sd.items():
+        full_param = sharded_param.full_tensor()
+        if is_rank_zero:
+            cpu_state_dict[param_name] = full_param.cpu()
```

Comment:
When saving QLoRA state dict, will we also upcast from NF4 here? (Obv not a concern for this PR)

Reply:
good question! I am open to discuss on 2 options in case you have preference in QLoRA PR
* torch.save(NF4) directly
* upcast and torch.save(full precision)

Reply:
I think the latter is preferable if possible, since NF4 is not really a universally-recognized dtype iiuc. This is what we do currently (see [here](https://github.com/pytorch/torchtune/blob/0c4056a6a55bfffd3cec7a99a13dc978cfe85963/torchtune/modules/common_utils.py#L15))

Reply:
the pointer is very helpful. now i know the torchtune way. will upcast then